src/buffer/out/ImageSlice.cpp

```diff
@@ -0,0 +1,245 @@
+        if (existingData)
+        {
+            // If there is existing data in the buffer, we need to copy it
+            // across to the appropriate position in the new buffer.
+            auto newPixelBuffer = std::vector<RGBQUAD>(bufferSize);
+            const auto newPixelOffset = (oldColumnBegin - _columnBegin) * _cellSize.width;
+            auto newIterator = std::next(newPixelBuffer.data(), newPixelOffset);
+            auto oldIterator = _pixelBuffer.data();
+            // Because widths are rounded up to multiples of 4, it's possible
+            // that the old width will extend past the right border of the new
+            // buffer, so the range that we copy must be clamped to fit.
+            const auto newPixelRange = std::min(oldPixelWidth, _pixelWidth - newPixelOffset);
+            for (auto i = 0; i < _cellSize.height; i++)
+            {
+                std::memcpy(newIterator, oldIterator, newPixelRange * sizeof(RGBQUAD));
+                std::advance(oldIterator, oldPixelWidth);
+                std::advance(newIterator, _pixelWidth);
+            }
+            _pixelBuffer = std::move(newPixelBuffer);
+        }
+        else
+        {
+            // Otherwise we just initialize the buffer to the correct size.
+            _pixelBuffer.resize(bufferSize);
+        }
+    }
+    const auto pixelOffset = (columnBegin - _columnBegin) * _cellSize.width;
+    return &til::at(_pixelBuffer, pixelOffset);
+}
+
+void ImageSlice::CopyBlock(const TextBuffer& srcBuffer, const til::rect srcRect, TextBuffer& dstBuffer, const til::rect dstRect)
+{
+    // If the top of the source is less than the top of the destination, we copy
+    // the rows from the bottom upwards, to avoid the possibility of the source
+    // being overwritten if it were to overlap the destination range.
+    if (srcRect.top < dstRect.top)
+    {
+        for (auto y = srcRect.height(); y-- > 0;)
+        {
+            const auto& srcRow = srcBuffer.GetRowByOffset(srcRect.top + y);
+            auto& dstRow = dstBuffer.GetMutableRowByOffset(dstRect.top + y);
+            CopyCells(srcRow, srcRect.left, dstRow, dstRect.left, dstRect.right);
+        }
+    }
+    else
+    {
+        for (auto y = 0; y < srcRect.height(); y++)
+        {
+            const auto& srcRow = srcBuffer.GetRowByOffset(srcRect.top + y);
+            auto& dstRow = dstBuffer.GetMutableRowByOffset(dstRect.top + y);
+            CopyCells(srcRow, srcRect.left, dstRow, dstRect.left, dstRect.right);
+        }
+    }
+}
+
+void ImageSlice::CopyRow(const ROW& srcRow, ROW& dstRow)
+{
+    const auto& srcSlice = srcRow.GetImageSlice();
+    auto& dstSlice = dstRow.GetMutableImageSlice();
+    dstSlice = srcSlice ? std::make_unique<ImageSlice>(*srcSlice) : nullptr;
+}
+
+void ImageSlice::CopyCells(const ROW& srcRow, const til::CoordType srcColumn, ROW& dstRow, const til::CoordType dstColumnBegin, const til::CoordType dstColumnEnd)

```

Comment:
I don't think that would be worth it as things are currently implemented. We already have the `ROW` references that are being passed to `CopyCells`, but if we were to go the `CopyBlock` route they would need to be looked up again. Along with everything else in `CopyBlock`, that's quite a lot more code that would need to be executed, and that's happening even when you aren't actually using sixel.

To be honest, I was expecting you'd eventually want to merge a lot of this `ImageSlice` code into the `TextBuffer` itself, so it would be easier to inline some of the checks that determine when sixel isn't in use, and that way we could lessen the impact on non-sixel users. My main reason for isolating it like this was to make it easer for me to maintain my fork, and also potentially easier to wrap in a feature flag if that becomes necessary. I know this is not the best long-term architecture.

Reply:
> To be honest, I was expecting you'd eventually want to merge a lot of this `ImageSlice` code into the `TextBuffer` itself [...]

Huh, of course... I didn't think of that yet, but that makes total sense!